### PR TITLE
Add non fatal errors to the engine

### PR DESCRIPTION
**Pull request by kerzhner**
When initially developing `go-nitro`, we adopted the methodology of panicking the engine when an error is encountered. This is approach works well during the development phase as we avoid introducing any errors during normal nitro node operation.

In production, a nitro node should panic infrequently. The panic scenario is when the node reaches an irrecoverable state. This PR moves in that direction by introducing a class of errors for which the node logs an error but does not panic. We then use the new non-fatal errors to fix https://github.com/statechannels/go-nitro/issues/1216.

That issue involves a duplicate, delayed message causing the store to look up a deleted consensus channel. While many different approaches exist to solving this particular issue (such as not deleting consensus channels or merging consensus channels and vanilla channels), this PR choses to classify the failure to fetch an objective as a non fatal error.

### CODE REVIEW

File: client/engine/engine.go [hunk 1]
```diff
@@ -721,3 +727,22 @@ func (e *Engine) recordMessageMetrics(message protocols.Message) {
 	e.metrics.RecordQueueLength(fmt.Sprintf("msg_payload_size,sender=%s,receiver=%s", e.store.GetAddress(), message.To), totalPayloadsSize)
 	e.metrics.RecordQueueLength(fmt.Sprintf("msg_size,sender=%s,receiver=%s", e.store.GetAddress(), message.To), len(raw))
 }
+
+func (e *Engine) checkError(err error) {
+	if err != nil {
+
+		e.logger.Err(err).Msgf("%s, error in run loop", e.store.GetAddress())
+
+		for _, nonFatalError := range nonFatalErrors {
```
Comment (geoknee) [inline client/engine/engine.go:736]: Would it make sense to invert this logic, and instead of keeping a list of nonFatalErrors, keep a list of fatalErrors and check against that? I think in production we want the default to be non-fatal. So far we have (I think) many examples of non fatal errors, but not so many (none?) fatal error.

File: client/engine/engine.go [hunk 2]
```diff
@@ -721,3 +727,22 @@ func (e *Engine) recordMessageMetrics(message protocols.Message) {
...
+			if errors.Is(err, nonFatalError) {
+				return
+			}
+		}
+
+		<-time.After(1000 * time.Millisecond) // We wait for a bit so the previous log line has time to complete
+		// TODO do not panic if in production.
```
Comment (geoknee) [inline client/engine/engine.go:743]: We could remove this TODO now, I think -- point being, we have pinched off errors that actually do require a panic in production.

File: client/engine/engine.go [hunk 3]
```diff
@@ -721,3 +727,22 @@ func (e *Engine) recordMessageMetrics(message protocols.Message) {
...
+		panic(err)
+		// TODO report errors back to the consuming application
```
Comment (geoknee) [inline client/engine/engine.go:745]: I think we could move this TODO to L738. More important to report errors that are (currently) being swallowed.

File: channel/consensus_channel/consensus_channel.go [hunk 4]
```diff
@@ -894,7 +894,10 @@ func (c *ConsensusChannel) Clone() *ConsensusChannel {
 	for i, p := range c.proposalQueue {
 		clonedProposalQueue[i] = p.Clone()
 	}
-	d := ConsensusChannel{c.MyIndex, c.fp.Clone(), c.Id, c.OnChainFunding.Clone(), c.current.clone(), clonedProposalQueue}
+	d := ConsensusChannel{
+		MyIndex: c.MyIndex, fp: c.fp.Clone(),
+		Id: c.Id, OnChainFunding: c.OnChainFunding.Clone(), current: c.current.clone(), proposalQueue: clonedProposalQueue,
+	}
```
Comment (geoknee) [inline channel/consensus_channel/consensus_channel.go:900]: Nit: not opposed to this change, but seems very unrelated to the PR. Also, if you are forming literals with named arguments, the order of the fields doesn't matter... so I'm not really understanding the benefit?

File: client/engine/engine.go (hunk 1 above)
Comment (kerzhner) [reply client/engine/engine.go:736]: I am open to either approach. My thinking was that for now we could still lean toward panicking to highlight unexpected errors

File: channel/consensus_channel/consensus_channel.go (hunk 4 above)
Comment (kerzhner) [reply channel/consensus_channel/consensus_channel.go:900]: The order of the fields of `consensus_channel` was changed to match that of channel. As a result, this struct construction started failing since the values were not named. Happy to move https://github.com/statechannels/go-nitro/pull/1239/commits/1ff9feb595192863d760c7d4b72a9adb6ee8a0fe to a separate PR if there is a strong preference.

File: client/engine/engine.go (hunk 1 above)
Comment (geoknee) [reply client/engine/engine.go:736]: Sure, OK. Seems safer this way.

File: client/engine/engine.go (hunk 3 above)
Comment (kerzhner) [reply client/engine/engine.go]: I chose to rephrase this todo and place it above the relevant line (the panic) ac9aaf36

Comment (geoknee) [reply client/engine/engine.go]: Maybe I'm confused, but I think we don't necessarily want to say "instead of panic" now. I believe we've got a model now where we want _some_ errors to result in a panic. Probably all errors should be sent to the API consumer / user.

**Review by geoknee — APPROVED** [review]
